engine/jobmaster/dm/dm_jobmaster.go

```diff
@@ -357,17 +357,18 @@ func (jm *JobMaster) cancel(ctx context.Context, code frameModel.WorkerStatusCod
 	}
 
 	if err := jm.taskManager.OperateTask(ctx, dmpkg.Deleting, nil, nil); err != nil {
-		return jm.Exit(ctx, status, err)
+		// would not recover again
+		return jm.Exit(ctx, framework.ExitReasonCancelled, err, "")
 	}
 	// wait all worker exit
 	jm.workerManager.SetNextCheckTime(time.Now())
 	for {
 		select {
 		case <-ctx.Done():
-			return jm.Exit(ctx, status, ctx.Err())
+			return jm.Exit(ctx, framework.ExitReasonCancelled, ctx.Err(), "")
 		case <-time.After(time.Second):
 			if jm.workerManager.allTombStone() {
-				return jm.Exit(ctx, status, err)
+				return jm.Exit(ctx, framework.ExitReasonCancelled, err, string(status.ExtBytes))
```

Comment:
The test failed because here. If code is frameModel.WorkerStatusFinished, the exit reason should be finished here.